metal/ringbuf.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#ifndef METAL__RINGBUF_H
+#define METAL__RINGBUF_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*!
+ * @brief The handle for a ring buffer
+ */
+struct metal_ringbuf {
+    /* _buf points at the base of the ring buffer memory */
+    void *_buf;
+    /* _start points at the ring buffer location containing the first value to
+     * be read by metal_ringbuf_get */
+    void *_start;
+    /* _end points at the first empty location after the buffer contents */
+    void *_end;
+    /* _capacity is the number of items the ring buffer can hold */
+    size_t _capacity;
+    /* _item_size is the size of each element in the ring buffer memory */
+    size_t _item_size;
```

Comment:
This does make the system more general, but I suspect the dominant use will be to buffer bytes. Would it be better to offer APIs that insert/remove multiple bytes and remove this field? It would also ensure that the amount of memory read/written by the queue/dequeue functions was part of that API, rather than being implicit from the configuration of the ring.